### PR TITLE
Pre publish security check

### PR DESCRIPTION
**Pull request by SCdF**
Allows for branches to exist with security issues, but prevents tags
being created with insecure package.

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

### CODE REVIEW

**Comment by SCdF** [comment]
Check out https://nodesecurity.io/opensource for more info.

This might not be the best place to do it, we can move it somewhere else. There is a grunt plugin that we could put into each `ci` task if we wanted. We could also use their github / travis integration that we can get for free with open source projects.

Consider this PR a discussion opener.

File: scripts/ci/push_to_staging.sh [hunk 1]
```diff
@@ -12,12 +12,24 @@
...
+# packages
+function nspcheck {
+    (echo 'Checking medic-webapp' && nsp check) ||
+    (echo 'Checking medic-api' && cd api && nsp check) ||
+    (echo 'Checking medic-sentinel' && cd sentinel && nsp check)
+}
+
 echo 'Building build for builds database...'
 if [[ -n "$TRAVIS_TAG" ]]; then
+    nspcheck
```
Comment (garethbowen) [inline scripts/ci/push_to_staging.sh]: Agreed. npm install then nspcheck right away IMO.

File: scripts/ci/push_to_staging.sh [hunk 2]
```diff
@@ -12,12 +12,24 @@
...
     node --stack_size=10000 `which kanso` push --minify \
             --id="$TRAVIS_TAG" \
             "$UPLOAD_URL"/_couch/builds
 else
+    nspcheck || true; # We're OK with this failing for branches
```
Comment (garethbowen) [inline scripts/ci/push_to_staging.sh]: Totally. Fail early. If it happens too often when there's no actual hack then we can rethink.

**Comment by garethbowen** [comment]
Great start. I agree with everything @mandric and @alxndrsn said. Check early, in every build, in every repo.

**Comment by SCdF** [comment]
> Yeah I would rather see separate tasks in each project and that would get included in the ci task in each project, that gets called by the travis ci in each project...

@mandric yeah me too. The reason I didn't go for that is that I wanted the check to occur before a publish. The reason being, that transitive dependencies could change. Thinking about that now though, that's mostly stupid, because package-lock should be solving that for us, and so it should be covered by each build. You *could* still release with medic-webapp having no issues but api and sentinel still having them. However, api and sentinel would be failing in travis, so it would be some kind of indication. We could also just call the grunt tasks in the pre-publish step *as well* just to double-check.

**Comment by SCdF** [comment]
FWIW all, I've noticed that NSP actually works without `node_modules`, which makes me wonder exactly what it's doing. It might be that it's not checking transitive dependencies, or it may be that it's just managing all of that server side. I'm going to:
 - Look into and make sure it's checking transitive dependencies as well
 - See if it has support for package-lock (documentation talks about shrinkwrap but not the newer package-lock)

**Comment by SCdF** [comment]
Closing in favour of doing it via grunt

**Pull request closed without merging**